// doker/lib/mixins/outlines.js
// Generated by CoffeeScript 1.9.3
(function() {
  module.exports = {
    addOutline: function(text, options) {
      var depth, outline, outlinesRoot, parent;
      depth = options.outline > 0 ? options.outline : 1;
      if (!this._root.data.Outlines) {
        this._root.data.Outlines = this.ref({
          Type: 'Outlines',
          Count: 0
        });
        this._root.data.PageMode = 'UseOutlines';
        this._outlines = [];
        this._lastForDepth = [this._root.data.Outlines];
      }
      outlinesRoot = this._root.data.Outlines;
      parent = this._lastForDepth[depth - 1] != null ? this._lastForDepth[depth - 1] : outlinesRoot;
      outline = this.ref({
        Title: new String(text),
        Parent: parent,
        Count: 0,
        Dest: [this.page.dictionary, 'XYZ', 0, this.page.height - this.y, 0]
      });
      if (parent !== outlinesRoot) {
        parent.data.Last = outline;
        if (parent.data.Count === 0) {
          parent.data.First = outline;
        }
        parent.data.Count++;
      }
      if (this._lastForDepth[depth] != null) {
        this._lastForDepth[depth].data.Next = outline;
        outline.data.Prev = this._lastForDepth[depth];
      }
      this._lastForDepth[depth] = outline;
      this._lastForDepth = this._lastForDepth.slice(0, +depth + 1 || 9e9);
      this._outlines.push(outline);
      if (outlinesRoot.data.Count === 0) {
        outlinesRoot.data.First = outline;
      }
      outlinesRoot.data.Last = outline;
      outlinesRoot.data.Count++;
      return this;
    },
    _endOutlines: function() {
      var i, j, len, outline, ref, results;
      if (this._root.data.Outlines != null) {
        this._root.data.Outlines.end();
        ref = this._outlines;
        results = [];
        for (i = j = 0, len = ref.length; j < len; i = ++j) {
          outline = ref[i];
          results.push(outline.end());
        }
        return results;
      }
    }
  };

}).call(this);
